File: 2021-07-21/poldo_game/sol/sol.py
#!/usr/bin/env python3
import sys

DEBUG = 0
#DEBUG = 1

N=int(input().strip())
colesterolo = list(map(int,input().strip().split())) # I panini (portate) sono numerati partendo da 0. Sugli indici pari sceglie Poldo cui spetta scegliere ad esempio sulla prima portata.
colesterolo.append(0)   #viene comodo un (N+1)-esimo panino fittizio con questo valore di colesterolo. 
P= [ [0] * (N+1) for j in range(N+1) ]  #  P[j][i] = numero di portate che verrebbero infine assaggiate se fossero interdette tutte le portate con valore di colesterolo <= colesterolo[j] e tutte le portate che precedono l'i-esima.
# Note: I panini (portate) sono numerati partendo da 0. Il primo a scegliere resta determinato da i%2 secondo la regola che sugli indici pari sceglie Poldo, cui ad esempio spetta di scegliere sulla prima portata.

for i in range(N-1,-1,-1):
    for j in range(N+1):
        P_i_se_non_mangio_i = P[j][i+1]
        if colesterolo[i] <= colesterolo[j]:
            P[j][i] = P_i_se_non_mangio_i
        else:
            P_i_se_mangio_i = 1 + P[i][i+1]
            if i%2==0:  #sceglie poldo
                P[j][i] = max(P_i_se_mangio_i, P_i_se_non_mangio_i)
            if i%2==1:  #sceglie il dottore
                P[j][i] = min(P_i_se_mangio_i, P_i_se_non_mangio_i)
print(P[N][0])
        
if DEBUG:
    for i in range(N):
        print(f"{P[N][i]}", end=" ")
    print("\n")
exit(0)

File: 2021-07-21/poldo_game/sol/sol.cpp
/**
 *  File: sol.cpp
 *  Soluzione di riferimento per poldo_game, quadratica nel tempo e lineare nello spazio
 *
 *  Autore: Romeo Rizzi, 2021-07-20
 *
 */

#include <cassert>
#include <cstdio>

const int DEBUG = 0;
//const int DEBUG = 1;
const int MAXN = 200;

typedef long long unsigned int llu;

int N;
int colesterolo[MAXN+1]; // I panini (portate) sono numerati partendo da 0. Sugli indici pari sceglie Poldo cui spetta scegliere ad esempio sulla prima portata. Viene però comodo un (N+1)-esimo panino fittizio con questo valore di colesterolo colesterolo[N]=0.
int P[MAXN+1][MAXN+1]; /* P[i][j] = numero di portate che verrebbero infine assaggiate se fossero interdette tutte le portate con valore di colesterolo <= colesterolo[j] e tutte le portate che precedono l'i-esima.
Note: I panini (portate) sono numerati partendo da 0. Il primo a scegliere resta determinato da i%2 secondo la regola che sugli indici pari sceglie Poldo, cui ad esempio spetta di scegliere sulla prima portata.*/

inline int my_min(int a, int b) { return (a<b) ? a : b; }
inline int my_max(int a, int b) { return (a>b) ? a : b; }

int main() {
  scanf("%d", &N);
  for(int i = 0; i < N; i++)
    scanf("%d", &colesterolo[i]);
  for(int i=N-1; i >= 0; i--) {
    for(int j=0; j<=N; j++) {
      int P_i_se_non_mangio_i = P[i+1][j];
      if(colesterolo[i] <= colesterolo[j])
	P[i][j] = P_i_se_non_mangio_i;
      else {      
	int P_i_se_mangio_i = 1 + P[i+1][i];
        if(i%2==0)  //sceglie poldo
	  P[i][j] = my_max(P_i_se_mangio_i, P_i_se_non_mangio_i);
        if(i%2==1)  //sceglie il dottore
	  P[i][j] = my_min(P_i_se_mangio_i, P_i_se_non_mangio_i);    
      }
    }
  }
  printf("%d\n", P[0][N]);
        
  if(DEBUG) {
    for(int i = 0; i < N; i++)
      printf("%d ", P[i][N]);
    printf("\n");
  }
  return 0;
}

